### ISSUE TITLE
Performance regression: very slow reduction in the presence of many module parameters

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#3102]
On stable and master I observe a severe performance regression on the test cases for #2402, dropping from instantaneous to 35s per test case to produce the expected error message
```
Incomplete pattern matching for test. Missing cases:
  test false
```
```
    Issue2402-1:                                       OK (34.79s)
    Issue2402-2:                                       OK (34.58s)
```
I am not sure whether 2.5.4 has this regression or just the current development version.

**Comment by nad** [comment]
This problem does not seem to affect Agda 2.5.4.

**Comment by andreasabel** [comment]
Somewhat mysteriously, the time seems to be spent before the actual coverage checking starts, in these lines:
https://github.com/agda/agda/blob/cb18777ff7c30f7df940db3188e7f7213bf15cc9/src/full/Agda/TypeChecking/Coverage.hs#L148-L160
The last patch affecting this region is 5 months old.

Time stats:
```
Total         15,500ms 
Miscellaneous     32ms 
Typing         7,768ms 
Coverage       7,700ms 
```

More investigation shows that the time apparently is spent in the first line, `TelV gamma a <- telView t`, and only takes this eternity if we have the import:
```agda
open import Agda.Builtin.Bool
-- data Bool : Set where  -- no problem without import
--   true false : Bool

module _ (A B C D E F G H I J K L M O P Q R S T U V W X Y Z
          -- A₁ B₁ C₁ D₁ E₁ F₁ G₁ H₁ I₁ J₁ K₁ L₁ M₁ O₁ P₁ Q₁ R₁ S₁ T₁ U₁ V₁ W₁ X₁ Y₁ Z₁ -- could not await this variant
          : Set) where

test : Bool → Bool
test true = false
test false = false
```

**Comment by andreasabel** [comment]
@UlfNorell @jespercockx Anyone tinkered with modules/parameters/imports/reduction in the development version after the release?
@nad: Maybe this can be tracked down by agda-bisect?

**Comment by UlfNorell** [comment]
I don't believe I have any commits not in the release. As far as I know the only new stuff is @jespercockx's Prop.

**Comment by andreasabel** [comment]
Bisection blames commit b9c7af0292375f74acf044f17fb6539d7f875561

    [ refactor ] Make univSort and univSort' monadic in HasOptions m

**Comment by nad** [comment]
> only takes this eternity if we have the import

Note that if you add `data [...]` before the top-level module header, then the header is no longer the header of a top-level module, but rather of an empty local module.

**Comment by andreasabel** [comment]
Ah, yes, silly me.

**Comment by nad** [comment]
I seem to recall that it was your idea to make it possible to omit the top-level module header. :)

**Comment by andreasabel** [comment]
My Idea was to be able to use `_` in the top level module header.  But then Ulf extended this to _drop it altogether_.

**Comment by UlfNorell** [comment]
I think you'll find you got that backwards: #953.

**Comment by andreasabel** [comment]
Memory is selective. :D

**Comment by andreasabel** [comment]
Back to the original issue.
@jespercockx : Could you document briefly for the future generations what the problem was and what fixed it?  Thanks.

**Comment by jespercockx** [comment]
Right. The performance regression was introduced because I turned `univSort'` (which computes the universe 'one level up' from a given sort) from a pure function into a `TCM` action (well, actually an `m` action for any `MonadReader TCState m`). The reason for this change was because when I introduced the new option `--omega-in-omega`, the reduction behaviour of `univSort'` depends on whether or not this option is enabled: if it is enabled, then `univSort' Inf = Just Inf`, otherwise it is `Nothing`.

The reason why this introduced a performance regression is actually a bit indirect: it meant substitution for sorts could no longer simplify sorts on-the-fly, since `applySubst` is a pure function. And (apparently) having the unreduced sorts everywhere caused a huge slowdown in some cases.

To fix the issue, I made `univSort'` again into a pure function, but with an extra argument of type `Maybe Sort` for the value of `univSort' Inf`. In most cases this argument is set to the correct value depending on the `--omega-in-omega` flag, but for `applySubst` it is always `Nothing`. This means `univSort Inf` will not be reduced on-the-fly during substitution, but at least all other `univSort`s will still be reduced right away.
